File: tool/fetch_fallback_puzzles.mjs
// Downloads a slice of the public Lichess puzzle DB (.zst) and curates a small
// set of approachable puzzles into assets/fallback_puzzles.json.
// Run: node tool/fetch_fallback_puzzles.mjs
import zlib from 'node:zlib';
import { writeFileSync, mkdirSync } from 'node:fs';

const URL = 'https://database.lichess.org/lichess_db_puzzle.csv.zst';
const WANT = 60;

const res = await fetch(URL, { headers: { Range: 'bytes=0-16000000' } });
const buf = Buffer.from(await res.arrayBuffer());
console.error(`downloaded ${buf.length} compressed bytes (HTTP ${res.status})`);

// Decompress as much as possible from the truncated slice. Do NOT call end()
// (which would error on the incomplete final frame and drop buffered output);
// instead write, let it emit, then stop.
const chunks = [];
const dec = zlib.createZstdDecompress();
dec.on('data', (d) => chunks.push(d));
dec.on('error', () => {});
dec.write(buf);
await new Promise((resolve) => setTimeout(resolve, 2500));
dec.destroy();
const text = Buffer.concat(chunks).toString('utf8');
const lines = text.split('\n');
console.error(`decompressed ${lines.length} lines`);

// CSV: PuzzleId,FEN,Moves,Rating,RatingDeviation,Popularity,NbPlays,Themes,GameUrl,OpeningTags
const picked = [];
for (let i = 1; i < lines.length && picked.length < WANT; i++) {
  const f = lines[i].split(',');
  if (f.length < 8) continue;
  const [id, fen, moves, rating, , popularity, nbPlays, themes] = f;
  const mv = moves.trim().split(' ');
  const r = parseInt(rating, 10);
  if (!id || !fen || mv.length < 2) continue;
  if (Number.isNaN(r) || r < 800 || r > 1700) continue;
  if (mv.length > 6) continue; // 1..2 solver moves, keep it short for an alarm
  if (parseInt(nbPlays, 10) < 30) continue;
  if (parseInt(popularity, 10) < 50) continue;
  if (themes.includes('veryLong') || themes.includes('long')) continue;
  picked.push({
    id,
    fen,
    moves: mv.join(' '),
    rating: r,
    themes: themes.trim().split(' ').slice(0, 6),
  });
}

mkdirSync('assets', { recursive: true });
writeFileSync('assets/fallback_puzzles.json', JSON.stringify(picked, null, 0));
console.error(`wrote ${picked.length} puzzles to assets/fallback_puzzles.json`);
if (picked.length) console.error('sample:', JSON.stringify(picked[0]));

File: tool/fetch_fallback_puzzles.py
"""Curate ~60 approachable puzzles from a slice of the public Lichess puzzle DB.
Run: python tool/fetch_fallback_puzzles.py"""
import io, json, os, urllib.request
import zstandard as zstd

URL = "https://database.lichess.org/lichess_db_puzzle.csv.zst"
WANT = 60

req = urllib.request.Request(URL, headers={"Range": "bytes=0-8000000"})
comp = urllib.request.urlopen(req).read()
print(f"downloaded {len(comp)} compressed bytes")

# Decompress incrementally; stop when the truncated slice runs out.
dctx = zstd.ZstdDecompressor()
reader = dctx.stream_reader(io.BytesIO(comp))
out = bytearray()
try:
    while len(out) < 60_000_000:
        chunk = reader.read(1 << 20)
        if not chunk:
            break
        out += chunk
except zstd.ZstdError:
    pass  # truncated final frame — keep what we decompressed
text = out.decode("utf-8", errors="ignore")
lines = text.split("\n")
print(f"decompressed {len(lines)} lines")

# CSV: PuzzleId,FEN,Moves,Rating,RatingDeviation,Popularity,NbPlays,Themes,GameUrl,OpeningTags
picked = []
for line in lines[1:]:
    f = line.split(",")
    if len(f) < 8:
        continue
    pid, fen, moves, rating, _dev, popularity, nbplays, themes = f[:8]
    mv = moves.strip().split(" ")
    try:
        r = int(rating); pop = int(popularity); plays = int(nbplays)
    except ValueError:
        continue
    if not pid or not fen or len(mv) < 2 or len(mv) > 6:
        continue
    if r < 800 or r > 1700 or plays < 30 or pop < 50:
        continue
    if "veryLong" in themes or "long" in themes:
        continue
    picked.append({
        "id": pid,
        "fen": fen,
        "moves": " ".join(mv),
        "rating": r,
        "themes": themes.strip().split(" ")[:6],
    })
    if len(picked) >= WANT:
        break

os.makedirs("assets", exist_ok=True)
with open("assets/fallback_puzzles.json", "w") as fp:
    json.dump(picked, fp)
print(f"wrote {len(picked)} puzzles")
if picked:
    print("sample:", json.dumps(picked[0]))
